**Clean Flights 2019**
<br/>Perform cleaning on the 2019 dataframe and then export a clean csv.

In [1]:
import numpy as np
import pandas as pd

In [2]:
import warnings
warnings.filterwarnings("ignore")

In [3]:
# increase display size to enable viewing of all data columns

pd.options.display.max_columns = 35

In [4]:
# read in the 2019 flights csv and ensure everything loads correctly

flights2019_df = pd.read_csv('flights2019.csv', index_col=0)
flights2019_df

DayOfWeek  FlightDate Reporting_Airline Tail_Number Origin  \
0                4  2019-08-01                DL      N354NW    ATL   
1                4  2019-08-01                DL      N320US    DFW   
2                4  2019-08-01                DL      N931DN    IAH   
3                4  2019-08-01                DL      N851DN    PDX   
4                4  2019-08-01                DL      N775DE    SLC   
...            ...         ...               ...         ...    ...   
7422032          5  2019-11-08                DL      N978AT    HOU   
7422033          5  2019-11-08                DL      N6716C    CHS   
7422034          5  2019-11-08                DL      N851DN    MSP   
7422035          5  2019-11-08                DL      N352NB    GEG   
7422036          5  2019-11-08                DL      N352NB    SLC   

                OriginCityName OriginState Dest           DestCityName  \
0                  Atlanta, GA          GA  DFW  Dallas/Fort Worth, TX   
1        Dallas/Fort Worth, TX          TX  ATL            Atlanta, GA   
2                  Houston, TX          TX  ATL            Atlanta, GA   
3                 Portland, OR          OR  SLC     Salt Lake City, UT   
4           Salt Lake City, UT          UT  PDX           Portland, OR   
...                        ...         ...  ...                    ...   
7422032            Houston, TX          TX  ATL            Atlanta, GA   
7422033         Charleston, SC          SC  ATL            Atlanta, GA   
7422034        Minneapolis, MN          MN  BWI          Baltimore, MD   
7422035            Spokane, WA          WA  SLC     Salt Lake City, UT   
7422036     Salt Lake City, UT          UT  GEG            Spokane, WA   

        DestState  CRSDepTime  DepTime  DepDelay  DepDel15  \
0              TX        1954   2025.0      31.0       1.0   
1              GA        1709   1706.0      -3.0       0.0   
2              GA        1749   1829.0      40.0       1.0   
3              UT        1310   1306.0      -4.0       0.0   
4              OR         829    820.0      -9.0       0.0   
...           ...         ...      ...       ...       ...   
7422032        GA        1324   1320.0      -4.0       0.0   
7422033        GA         700    719.0      19.0       1.0   
7422034        MD        1800   1754.0      -6.0       0.0   
7422035        UT        1015   1009.0      -6.0       0.0   
7422036        WA         844    836.0      -8.0       0.0   

         DepartureDelayGroups  TaxiOut  WheelsOff  Cancelled CancellationCode  \
0                         2.0     17.0     2042.0          0              NaN   
1                        -1.0     19.0     1725.0          0              NaN   
2                         2.0     12.0     1841.0          0              NaN   
3                        -1.0     13.0     1319.0          0              NaN   
4                        -1.0     16.0      836.0          0              NaN   
...                       ...      ...        ...        ...              ...   
7422032                  -1.0     14.0     1334.0          0              NaN   
7422033                   1.0     10.0      729.0          0              NaN   
7422034                  -1.0     22.0     1816.0          0              NaN   
7422035                  -1.0     11.0     1020.0          0              NaN   
7422036                  -1.0     14.0      850.0          0              NaN   

         Distance  DistanceGroup  CarrierDelay  WeatherDelay  NASDelay  \
0             731              3           1.0           0.0       0.0   
1             731              3           NaN           NaN       NaN   
2             689              3          17.0           0.0       0.0   
3             630              3           NaN           NaN       NaN   
4             630              3           NaN           NaN       NaN   
...           ...            ...           ...           ...       ...   
7422032       696              

In [5]:
# summary statistics for the dataframe

flights2019_df.describe().transpose()

count         mean         std   min     25%  \
DayOfWeek             7422037.0     3.937155    1.995814   1.0     2.0   
CRSDepTime            7422037.0  1330.258507  492.985466   1.0   913.0   
DepTime               7291951.0  1334.606311  507.199818   1.0   914.0   
DepDelay              7291927.0    10.923267   48.959640 -82.0    -5.0   
DepDel15              7291927.0     0.186804    0.389754   0.0     0.0   
DepartureDelayGroups  7291927.0     0.080959    2.276868  -2.0    -1.0   
TaxiOut               7288060.0    17.389349   10.003982   1.0    11.0   
WheelsOff             7288060.0  1358.264390  508.810907   1.0   930.0   
Cancelled             7422037.0     0.018179    0.133598   0.0     0.0   
Distance              7422037.0   800.544232  592.511847  31.0   369.0   
DistanceGroup         7422037.0     3.676005    2.332646   1.0     2.0   
CarrierDelay          1389253.0    21.128592   66.104645   0.0     0.0   
WeatherDelay          1389253.0     3.802404   32.362973   0.0     0.0   
NASDelay              1389253.0    16.587946   39.664395   0.0     0.0   
SecurityDelay         1389253.0     0.096083    3.358346   0.0     0.0   
LateAircraftDelay     1389253.0    27.407479   53.456198   0.0     0.0   
CRSArrTime            7422037.0  1485.789553  521.465892   1.0  1100.0   
ArrTime               7284391.0  1462.955325  542.454249   1.0  1046.0   
ArrDelay              7268232.0     5.414849   51.066656 -99.0   -15.0   

                         50%     75%     max  
DayOfWeek                4.0     6.0     7.0  
CRSDepTime            1321.0  1736.0  2359.0  
DepTime               1327.0  1746.0  2400.0  
DepDelay                -2.0     7.0  2710.0  
DepDel15                 0.0     0.0     1.0  
DepartureDelayGroups    -1.0     0.0    12.0  
TaxiOut                 15.0    20.0   227.0  
WheelsOff             1340.0  1801.0  2400.0  
Cancelled                0.0     0.0     1.0  
Distance               640.0  1034.0  5095.0  
DistanceGroup            3.0     5.0    11.0  
CarrierDelay             0.0    18.0  2695.0  
WeatherDelay             0.0     0.0  1847.0  
NASDelay                 2.0    20.0  1741.0  
SecurityDelay            0.0     0.0  1078.0  
LateAircraftDelay        3.0    33.0  2206.0  
CRSArrTime            1515.0  1921.0  2400.0  
ArrTime               1504.0  1917.0  2400.0  
ArrDelay                -6.0     7.0  2695.0

In [6]:
# check to see if there any NaN values in the data

flights2019_df.isna().sum()

DayOfWeek                     0
FlightDate                    0
Reporting_Airline             0
Tail_Number               17837
Origin                        0
OriginCityName                0
OriginState                   0
Dest                          0
DestCityName                  0
DestState                     0
CRSDepTime                    0
DepTime                  130086
DepDelay                 130110
DepDel15                 130110
DepartureDelayGroups     130110
TaxiOut                  133977
WheelsOff                133977
Cancelled                     0
CancellationCode        7287112
Distance                      0
DistanceGroup                 0
CarrierDelay            6032784
WeatherDelay            6032784
NASDelay                6032784
SecurityDelay           6032784
LateAircraftDelay       6032784
CRSArrTime                    0
ArrTime                  137646
ArrDelay                 153805
dtype: int64

In [7]:
# list of all the columns in the data dataframe with missing values and the percent that's missing

round(np.mean(flights2019_df[flights2019_df.columns[flights2019_df.isnull().any()]].isna(), axis = 0) * 100,2)

Tail_Number              0.24
DepTime                  1.75
DepDelay                 1.75
DepDel15                 1.75
DepartureDelayGroups     1.75
TaxiOut                  1.81
WheelsOff                1.81
CancellationCode        98.18
CarrierDelay            81.28
WeatherDelay            81.28
NASDelay                81.28
SecurityDelay           81.28
LateAircraftDelay       81.28
ArrTime                  1.85
ArrDelay                 2.07
dtype: float64

In [8]:
# show flights that are missing the tail number and were cancelled

flights2019_df[(flights2019_df['Cancelled'] == 1) & (flights2019_df['Tail_Number'].isna())]

DayOfWeek  FlightDate Reporting_Airline Tail_Number Origin  \
40202            6  2019-08-24                DL         NaN    ORD   
93619            6  2019-08-03                YV         NaN    ELP   
93772            3  2019-08-07                YV         NaN    MRY   
94611            2  2019-08-20                YV         NaN    DFW   
94642            2  2019-08-20                YV         NaN    ICT   
...            ...         ...               ...         ...    ...   
7212228          2  2019-11-26                YV         NaN    DFW   
7212229          3  2019-11-27                YV         NaN    DFW   
7221600          2  2019-11-05                YV         NaN    JAN   
7222311          3  2019-11-06                YV         NaN    FSD   
7222332          3  2019-11-27                YV         NaN    FSD   

                OriginCityName OriginState Dest           DestCityName  \
40202              Chicago, IL          IL  LGA           New York, NY   
93619              El Paso, TX          TX  PHX            Phoenix, AZ   
93772             Monterey, CA          CA  PHX            Phoenix, AZ   
94611    Dallas/Fort Worth, TX          TX  ICT            Wichita, KS   
94642              Wichita, KS          KS  DFW  Dallas/Fort Worth, TX   
...                        ...         ...  ...                    ...   
7212228  Dallas/Fort Worth, TX          TX  CRP     Corpus Christi, TX   
7212229  Dallas/Fort Worth, TX          TX  CRP     Corpus Christi, TX   
7221600  Jackson/Vicksburg, MS          MS  DFW  Dallas/Fort Worth, TX   
7222311        Sioux Falls, SD          SD  DFW  Dallas/Fort Worth, TX   
7222332        Sioux Falls, SD          SD  DFW  Dallas/Fort Worth, TX   

        DestState  CRSDepTime  DepTime  DepDelay  DepDel15  \
40202          NY        1145      NaN       NaN       NaN   
93619          AZ         829      NaN       NaN       NaN   
93772          AZ         608      NaN       NaN       NaN   
94611          KS        1228      NaN       NaN       NaN   
94642          TX        1423      NaN       NaN       NaN   
...           ...         ...      ...       ...       ...   
7212228        TX        1430      NaN       NaN       NaN   
7212229        TX        1430      NaN       NaN       NaN   
7221600        TX        1625      NaN       NaN       NaN   
7222311        TX         700      NaN       NaN       NaN   
7222332        TX         700      NaN       NaN       NaN   

         DepartureDelayGroups  TaxiOut  WheelsOff  Cancelled CancellationCode  \
40202                     NaN      NaN        NaN          1                A   
93619                     NaN      NaN        NaN          1                A   
93772                     NaN      NaN        NaN          1                A   
94611                     NaN      NaN        NaN          1                A   
94642                     NaN      NaN        NaN          1                A   
...                       ...      ...        ...        ...              ...   
7212228                   NaN      NaN        NaN          1                A   
7212229                   NaN      NaN        NaN          1                A   
7221600                   NaN      NaN        NaN          1                A   
7222311                   NaN      NaN        NaN          1                A   
7222332                   NaN      NaN        NaN          1                B   

         Distance  DistanceGroup  CarrierDelay  WeatherDelay  NASDelay  \
40202         733              3           NaN           NaN       NaN   
93619         347              2           NaN           NaN       NaN   
93772         598              3           NaN           NaN       NaN   
94611         328              2           NaN           NaN       NaN   
94642         328              2           NaN           NaN       NaN   
...           ...            ...           ...           ...       ...   
7212228       354              

In [9]:
# fill in 'unknown' for the NaN tail numbers
# these are all flights that were cancelled, and we already have a small number of cancelled flights, so don't want to delete them

flights2019_df['Tail_Number'].fillna('unknown', inplace=True)
flights2019_df.shape

(7422037, 29)

In [10]:
# check to ensure that tail number was filled in appropriately

flights2019_df.isna().sum()

DayOfWeek                     0
FlightDate                    0
Reporting_Airline             0
Tail_Number                   0
Origin                        0
OriginCityName                0
OriginState                   0
Dest                          0
DestCityName                  0
DestState                     0
CRSDepTime                    0
DepTime                  130086
DepDelay                 130110
DepDel15                 130110
DepartureDelayGroups     130110
TaxiOut                  133977
WheelsOff                133977
Cancelled                     0
CancellationCode        7287112
Distance                      0
DistanceGroup                 0
CarrierDelay            6032784
WeatherDelay            6032784
NASDelay                6032784
SecurityDelay           6032784
LateAircraftDelay       6032784
CRSArrTime                    0
ArrTime                  137646
ArrDelay                 153805
dtype: int64

In [11]:
# check to see how many cancelled flights there were in 2019

flights2019_df[flights2019_df['Cancelled'] == 1].shape

(134925, 29)

134,925 cancelled flights, but 130,086 with no departure time -- there are 4,839 (see below) that need updating.

In [12]:
# look at the flights that are both cancelled and have a departure time

flights2019_df[(flights2019_df['Cancelled'] ==1) & (flights2019_df['DepTime'])]

DayOfWeek  FlightDate Reporting_Airline Tail_Number Origin  \
49969            3  2019-08-07                DL      N358NB    DTW   
50470            3  2019-08-07                DL      N984DL    PNS   
64816            3  2019-08-28                DL      N685DA    JFK   
64885            3  2019-08-28                DL      N551NW    ATL   
64961            3  2019-08-28                DL      N690DL    ATL   
...            ...         ...               ...         ...    ...   
7365137          1  2019-11-11                AA      N301PA    DFW   
7368189          2  2019-11-12                AA      N198AA    DFW   
7370399          5  2019-11-01                B6      N273JB    JFK   
7373537          2  2019-11-05                B6      N283JB    ORH   
7386077          1  2019-11-11                B6      N206JB    DTW   

                OriginCityName OriginState Dest          DestCityName  \
49969              Detroit, MI          MI  PHL      Philadelphia, PA   
50470            Pensacola, FL          FL  ATL           Atlanta, GA   
64816             New York, NY          NY  STT  Charlotte Amalie, VI   
64885              Atlanta, GA          GA  STT  Charlotte Amalie, VI   
64961              Atlanta, GA          GA  STT  Charlotte Amalie, VI   
...                        ...         ...  ...                   ...   
7365137  Dallas/Fort Worth, TX          TX  DEN            Denver, CO   
7368189  Dallas/Fort Worth, TX          TX  PHL      Philadelphia, PA   
7370399           New York, NY          NY  BTV        Burlington, VT   
7373537          Worcester, MA          MA  MCO           Orlando, FL   
7386077            Detroit, MI          MI  BOS            Boston, MA   

        DestState  CRSDepTime  DepTime  DepDelay  DepDel15  \
49969          PA        1414   1408.0      -6.0       0.0   
50470          GA        1559   1602.0       3.0       0.0   
64816          VI         900    903.0       3.0       0.0   
64885          VI         940   1012.0      32.0       1.0   
64961          VI        1107   1104.0      -3.0       0.0   
...           ...         ...      ...       ...       ...   
7365137        CO        1900   1858.0      -2.0       0.0   
7368189        PA         849    904.0      15.0       1.0   
7370399        VT         735    741.0       6.0       0.0   
7373537        FL        1608   2132.0     324.0       1.0   
7386077        MA        1858   2243.0     225.0       1.0   

         DepartureDelayGroups  TaxiOut  WheelsOff  Cancelled CancellationCode  \
49969                    -1.0     19.0     1427.0          1                C   
50470                     0.0     10.0     1612.0          1                A   
64816                     0.0     84.0     1027.0          1                B   
64885                     2.0     21.0     1033.0          1                B   
64961                    -1.0     11.0     1115.0          1                B   
...                       ...      ...        ...        ...              ...   
7365137                  -1.0      NaN        NaN          1                A   
7368189                   1.0      NaN        NaN          1                A   
7370399                   0.0     11.0      752.0          1                B   
7373537                  12.0      NaN        NaN          1                A   
7386077                  12.0      NaN        NaN          1                A   

         Distance  DistanceGroup  CarrierDelay  WeatherDelay  NASDelay  \
49969         453              2           NaN           NaN       NaN   
50470         271              2           NaN           NaN       NaN   
64816        1623              7           NaN           NaN       NaN   
64885        1599              7           NaN           NaN       NaN   
64961        1599              7           NaN           NaN       NaN   
...           ...            ...           ...           ...       ...   
7365137       641              3           

These don't have arrival times, so it is safe to assume that they were actually cancelled, and should not have a departure time.

In [13]:
# change the departure time to 0 for all flights that are cancelled and have a depature time
# and check that there are now no flights with this combination

flights2019_df['DepTime'][(flights2019_df['Cancelled'] == 1) & (flights2019_df['DepTime'])] = 0
flights2019_df[(flights2019_df['Cancelled'] ==1) & (flights2019_df['DepTime'])]

Empty DataFrame
Columns: [DayOfWeek, FlightDate, Reporting_Airline, Tail_Number, Origin, OriginCityName, OriginState, Dest, DestCityName, DestState, CRSDepTime, DepTime, DepDelay, DepDel15, DepartureDelayGroups, TaxiOut, WheelsOff, Cancelled, CancellationCode, Distance, DistanceGroup, CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay, CRSArrTime, ArrTime, ArrDelay]
Index: []

In [14]:
# check for flights that have a depature time greater than 0, but no departure delay

flights2019_df[(flights2019_df['DepTime'] > 0 ) & (flights2019_df['DepDelay'].isna())]

Empty DataFrame
Columns: [DayOfWeek, FlightDate, Reporting_Airline, Tail_Number, Origin, OriginCityName, OriginState, Dest, DestCityName, DestState, CRSDepTime, DepTime, DepDelay, DepDel15, DepartureDelayGroups, TaxiOut, WheelsOff, Cancelled, CancellationCode, Distance, DistanceGroup, CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay, CRSArrTime, ArrTime, ArrDelay]
Index: []

In [15]:
# review remaining flights with NaN values in the DepDelay column

flights2019_df[(flights2019_df['DepDelay'].isna())]

DayOfWeek  FlightDate Reporting_Airline Tail_Number Origin  \
893              4  2019-08-01                DL      N103DU    EWR   
1126             4  2019-08-01                DL      N334DN    LGA   
1289             4  2019-08-01                DL      N382DN    LGA   
40202            6  2019-08-24                DL     unknown    ORD   
49419            3  2019-08-07                DL      N374DX    ATL   
...            ...         ...               ...         ...    ...   
7390903          7  2019-11-17                B6      N645JB    RSW   
7402106          2  2019-11-12                DL      N968AT    SYR   
7402423          2  2019-11-12                DL      N320NB    ALB   
7402676          2  2019-11-12                DL      N355DN    MKE   
7402887          2  2019-11-12                DL      N928AT    BNA   

         OriginCityName OriginState Dest  DestCityName DestState  CRSDepTime  \
893          Newark, NJ          NJ  DTW   Detroit, MI        MI         600   
1126       New York, NY          NY  MIA     Miami, FL        FL         700   
1289       New York, NY          NY  ATL   Atlanta, GA        GA         959   
40202       Chicago, IL          IL  LGA  New York, NY        NY        1145   
49419       Atlanta, GA          GA  BOS    Boston, MA        MA        1829   
...                 ...         ...  ...           ...       ...         ...   
7390903  Fort Myers, FL          FL  BOS    Boston, MA        MA        1842   
7402106    Syracuse, NY          NY  DTW   Detroit, MI        MI         600   
7402423      Albany, NY          NY  DTW   Detroit, MI        MI         555   
7402676   Milwaukee, WI          WI  DTW   Detroit, MI        MI         632   
7402887   Nashville, TN          TN  DTW   Detroit, MI        MI         635   

         DepTime  DepDelay  DepDel15  DepartureDelayGroups  TaxiOut  \
893          NaN       NaN       NaN                   NaN      NaN   
1126         NaN       NaN       NaN                   NaN      NaN   
1289         NaN       NaN       NaN                   NaN      NaN   
40202        NaN       NaN       NaN                   NaN      NaN   
49419        NaN       NaN       NaN                   NaN      NaN   
...          ...       ...       ...                   ...      ...   
7390903      NaN       NaN       NaN                   NaN      NaN   
7402106      NaN       NaN       NaN                   NaN      NaN   
7402423      NaN       NaN       NaN                   NaN      NaN   
7402676      NaN       NaN       NaN                   NaN      NaN   
7402887      NaN       NaN       NaN                   NaN      NaN   

         WheelsOff  Cancelled CancellationCode  Distance  DistanceGroup  \
893            NaN          1                A       488              2   
1126           NaN          1                A      1096              5   
1289           NaN          1                A       762              4   
40202          NaN          1                A       733              3   
49419          NaN          1                C       946              4   
...            ...        ...              ...       ...            ...   
7390903        NaN          1                A      1249              5   
7402106        NaN          1                A       374              2   
7402423        NaN          1                A       489              2   
7402676        NaN          1                A       237              1   
7402887        NaN          1                A       456              2   

         CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
893               NaN           NaN       NaN            NaN   
1126              NaN           NaN       NaN            NaN   
1289              NaN           NaN       NaN            NaN   
40202             NaN           NaN       NaN            NaN   
49419             NaN           NaN       NaN            NaN   
...               ...           ...       ...            

In [16]:
# fill in 0 for the remaining NaN values in the DepTime, DepDelay, DepDel15 and DepartureDelayGroups columns

flights2019_df['DepTime'].fillna(0, inplace=True)
flights2019_df['DepDelay'].fillna(0, inplace=True)
flights2019_df['DepDel15'].fillna(0, inplace=True)
flights2019_df['DepartureDelayGroups'].fillna(0, inplace=True)
flights2019_df.isna().sum()

DayOfWeek                     0
FlightDate                    0
Reporting_Airline             0
Tail_Number                   0
Origin                        0
OriginCityName                0
OriginState                   0
Dest                          0
DestCityName                  0
DestState                     0
CRSDepTime                    0
DepTime                       0
DepDelay                      0
DepDel15                      0
DepartureDelayGroups          0
TaxiOut                  133977
WheelsOff                133977
Cancelled                     0
CancellationCode        7287112
Distance                      0
DistanceGroup                 0
CarrierDelay            6032784
WeatherDelay            6032784
NASDelay                6032784
SecurityDelay           6032784
LateAircraftDelay       6032784
CRSArrTime                    0
ArrTime                  137646
ArrDelay                 153805
dtype: int64

In [17]:
# look at the flights that are both cancelled and have a departure time

flights2019_df[(flights2019_df['Cancelled'] ==1) & (flights2019_df['ArrTime'])]

Empty DataFrame
Columns: [DayOfWeek, FlightDate, Reporting_Airline, Tail_Number, Origin, OriginCityName, OriginState, Dest, DestCityName, DestState, CRSDepTime, DepTime, DepDelay, DepDel15, DepartureDelayGroups, TaxiOut, WheelsOff, Cancelled, CancellationCode, Distance, DistanceGroup, CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay, CRSArrTime, ArrTime, ArrDelay]
Index: []

In [18]:
# inspect the columns with NaN values for arrival time that are not cancelled

flights2019_df[(flights2019_df['Cancelled'] == 0) & (flights2019_df['ArrTime'].isna())]

DayOfWeek  FlightDate Reporting_Airline Tail_Number Origin  \
47884            2  2019-08-06                DL      N348NB    RSW   
50386            3  2019-08-07                DL      N384DN    MSP   
51567            3  2019-08-07                DL      N317US    DTW   
52390            3  2019-08-07                DL      N354DN    FLL   
54178            4  2019-08-08                DL      N112DU    DFW   
...            ...         ...               ...         ...    ...   
7365406          6  2019-11-16                AA      N817NN    ORD   
7370272          5  2019-11-01                B6      N353JB    MCO   
7376357          6  2019-11-09                B6      N294JB    TPA   
7376455          6  2019-11-09                B6      N304JB    FLL   
7405755          3  2019-11-13                DL      N356NW    IAH   

                OriginCityName OriginState Dest      DestCityName DestState  \
47884           Fort Myers, FL          FL  LGA      New York, NY        NY   
50386          Minneapolis, MN          MN  BWI     Baltimore, MD        MD   
51567              Detroit, MI          MI  LGA      New York, NY        NY   
52390      Fort Lauderdale, FL          FL  LGA      New York, NY        NY   
54178    Dallas/Fort Worth, TX          TX  LGA      New York, NY        NY   
...                        ...         ...  ...               ...       ...   
7365406            Chicago, IL          IL  DCA    Washington, DC        VA   
7370272            Orlando, FL          FL  HPN  White Plains, NY        NY   
7376357              Tampa, FL          FL  HPN  White Plains, NY        NY   
7376455    Fort Lauderdale, FL          FL  HPN  White Plains, NY        NY   
7405755            Houston, TX          TX  LGA      New York, NY        NY   

         CRSDepTime  DepTime  DepDelay  DepDel15  DepartureDelayGroups  \
47884          1959   2328.0     209.0       1.0                  12.0   
50386          1253   1251.0      -2.0       0.0                  -1.0   
51567          1359     48.0     649.0       1.0                  12.0   
52390          1144   1147.0       3.0       0.0                   0.0   
54178          1405   1459.0      54.0       1.0                   3.0   
...             ...      ...       ...       ...                   ...   
7365406         707    703.0      -4.0       0.0                  -1.0   
7370272        1700   2048.0     228.0       1.0                  12.0   
7376357        1812   2048.0     156.0       1.0                  10.0   
7376455        1808   2103.0     175.0       1.0                  11.0   
7405755        1909   2236.0     207.0       1.0                  12.0   

         TaxiOut  WheelsOff  Cancelled CancellationCode  Distance  \
47884       47.0       15.0          0              NaN      1080   
50386       17.0     1308.0          0              NaN       936   
51567       14.0      102.0          0              NaN       502   
52390       24.0     1211.0          0              NaN      1076   
54178       21.0     1520.0          0              NaN      1389   
...          ...        ...        ...              ...       ...   
7365406     29.0      732.0          0              NaN       612   
7370272     18.0     2106.0          0              NaN       972   
7376357     15.0     2103.0          0              NaN      1032   
7376455     19.0     2122.0          0              NaN      1097   
7405755     19.0     2255.0          0              NaN      1416   

         DistanceGroup  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
47884                5           NaN           NaN       NaN            NaN   
50386                4           NaN           NaN       NaN            NaN   
51567                3           NaN           NaN       NaN            NaN   
52390                5           NaN           NaN       NaN            NaN   
54178                6           NaN           NaN       NaN            NaN   
...            

In [19]:
# there are 2,721 rows with flights that are not cancelled, and have a departure time, but no arrival time
# delete these rows

flights2019_df = flights2019_df.drop(flights2019_df[(flights2019_df['Cancelled'] == 0) & (flights2019_df['ArrTime'].isna())].index)
flights2019_df[(flights2019_df['Cancelled'] == 0) & (flights2019_df['ArrTime'].isna())]

Empty DataFrame
Columns: [DayOfWeek, FlightDate, Reporting_Airline, Tail_Number, Origin, OriginCityName, OriginState, Dest, DestCityName, DestState, CRSDepTime, DepTime, DepDelay, DepDel15, DepartureDelayGroups, TaxiOut, WheelsOff, Cancelled, CancellationCode, Distance, DistanceGroup, CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay, CRSArrTime, ArrTime, ArrDelay]
Index: []

In [20]:
# replace NaN with 0 for all flights that are cancelled 
# and check that there are now no flights with this combination

flights2019_df['ArrTime'][(flights2019_df['Cancelled'] == 1) & (flights2019_df['ArrTime'].isna())] = 0
flights2019_df['ArrDelay'][(flights2019_df['Cancelled'] == 1) & (flights2019_df['ArrDelay'].isna())] = 0
flights2019_df[(flights2019_df['Cancelled'] == 1) & (flights2019_df['ArrTime'])]

Empty DataFrame
Columns: [DayOfWeek, FlightDate, Reporting_Airline, Tail_Number, Origin, OriginCityName, OriginState, Dest, DestCityName, DestState, CRSDepTime, DepTime, DepDelay, DepDel15, DepartureDelayGroups, TaxiOut, WheelsOff, Cancelled, CancellationCode, Distance, DistanceGroup, CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay, CRSArrTime, ArrTime, ArrDelay]
Index: []

In [21]:
# check the status of NaNs

flights2019_df.isna().sum()

DayOfWeek                     0
FlightDate                    0
Reporting_Airline             0
Tail_Number                   0
Origin                        0
OriginCityName                0
OriginState                   0
Dest                          0
DestCityName                  0
DestState                     0
CRSDepTime                    0
DepTime                       0
DepDelay                      0
DepDel15                      0
DepartureDelayGroups          0
TaxiOut                  133977
WheelsOff                133977
Cancelled                     0
CancellationCode        7284391
Distance                      0
DistanceGroup                 0
CarrierDelay            6030063
WeatherDelay            6030063
NASDelay                6030063
SecurityDelay           6030063
LateAircraftDelay       6030063
CRSArrTime                    0
ArrTime                       0
ArrDelay                  16159
dtype: int64

In [22]:
# look at the remaining flights with NaN values for ArrDelay

flights2019_df[(flights2019_df['ArrDelay'].isna()) & (flights2019_df['ArrTime'])]

DayOfWeek  FlightDate Reporting_Airline Tail_Number Origin  \
16               4  2019-08-01                DL      N928AT    ATL   
317              4  2019-08-01                DL      N684DA    SLC   
975              4  2019-08-01                DL      N967DL    ATL   
1007             4  2019-08-01                DL      N922AT    MSN   
1201             4  2019-08-01                DL      N988AT    DTW   
...            ...         ...               ...         ...    ...   
7417696          4  2019-11-07                DL      N920DN    ATL   
7420086          4  2019-11-07                DL      N966DL    ATL   
7420500          5  2019-11-08                DL      N121DU    FAI   
7420979          5  2019-11-08                DL      N669DN    MCO   
7421560          5  2019-11-08                DL       N3756    LAX   

             OriginCityName OriginState Dest            DestCityName  \
16              Atlanta, GA          GA  GSP               Greer, SC   
317      Salt Lake City, UT          UT  ATL             Atlanta, GA   
975             Atlanta, GA          GA  SRQ  Sarasota/Bradenton, FL   
1007            Madison, WI          WI  ATL             Atlanta, GA   
1201            Detroit, MI          MI  CLT           Charlotte, NC   
...                     ...         ...  ...                     ...   
7417696         Atlanta, GA          GA  ORD             Chicago, IL   
7420086         Atlanta, GA          GA  JAN   Jackson/Vicksburg, MS   
7420500       Fairbanks, AK          AK  SEA             Seattle, WA   
7420979         Orlando, FL          FL  SEA             Seattle, WA   
7421560     Los Angeles, CA          CA  SEA             Seattle, WA   

        DestState  CRSDepTime  DepTime  DepDelay  DepDel15  \
16             SC        1654   1702.0       8.0       0.0   
317            GA        1105   1115.0      10.0       0.0   
975            FL        1503   1526.0      23.0       1.0   
1007           GA        1523   1520.0      -3.0       0.0   
1201           NC        1730   1729.0      -1.0       0.0   
...           ...         ...      ...       ...       ...   
7417696        IL        1111   1110.0      -1.0       0.0   
7420086        MS        2255   2306.0      11.0       0.0   
7420500        WA          55    322.0     147.0       1.0   
7420979        WA         612    609.0      -3.0       0.0   
7421560        WA         600    557.0      -3.0       0.0   

         DepartureDelayGroups  TaxiOut  WheelsOff  Cancelled CancellationCode  \
16                        0.0     37.0     1739.0          0              NaN   
317                       0.0     13.0     1128.0          0              NaN   
975                       1.0     36.0     1602.0          0              NaN   
1007                     -1.0     10.0     1530.0          0              NaN   
1201                     -1.0     10.0     1739.0          0              NaN   
...                       ...      ...        ...        ...              ...   
7417696                  -1.0     14.0     1124.0          0              NaN   
7420086                   0.0     13.0     2319.0          0              NaN   
7420500                   9.0     24.0      346.0          0              NaN   
7420979                  -1.0     15.0      624.0          0              NaN   
7421560                  -1.0     20.0      617.0          0              NaN   

         Distance  DistanceGroup  CarrierDelay  WeatherDelay  NASDelay  \
16            153              1           NaN           NaN       NaN   
317          1590              7           NaN           NaN       NaN   
975           444              2           NaN           NaN       NaN   
1007          707              3           NaN           NaN       NaN   
1201          500              3           NaN           NaN       NaN   
...           ...            ...           ...           ...       ...   
7417696       606              3           NaN         

In [23]:
# fill in these NaN values with the difference between the expected arrival time (CRSArrTime) and the actual arrival time (ArrTime)

flights2019_df['ArrDelay'].fillna((flights2019_df['ArrTime'] - flights2019_df['CRSArrTime']), inplace=True)
flights2019_df[(flights2019_df['ArrDelay'].isna()) & (flights2019_df['ArrTime'])]

Empty DataFrame
Columns: [DayOfWeek, FlightDate, Reporting_Airline, Tail_Number, Origin, OriginCityName, OriginState, Dest, DestCityName, DestState, CRSDepTime, DepTime, DepDelay, DepDel15, DepartureDelayGroups, TaxiOut, WheelsOff, Cancelled, CancellationCode, Distance, DistanceGroup, CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay, CRSArrTime, ArrTime, ArrDelay]
Index: []

In [24]:
# check the status of NaNs

flights2019_df.isna().sum()

DayOfWeek                     0
FlightDate                    0
Reporting_Airline             0
Tail_Number                   0
Origin                        0
OriginCityName                0
OriginState                   0
Dest                          0
DestCityName                  0
DestState                     0
CRSDepTime                    0
DepTime                       0
DepDelay                      0
DepDel15                      0
DepartureDelayGroups          0
TaxiOut                  133977
WheelsOff                133977
Cancelled                     0
CancellationCode        7284391
Distance                      0
DistanceGroup                 0
CarrierDelay            6030063
WeatherDelay            6030063
NASDelay                6030063
SecurityDelay           6030063
LateAircraftDelay       6030063
CRSArrTime                    0
ArrTime                       0
ArrDelay                      0
dtype: int64

In [25]:
# fill in the TaxiOut column with the median value 

flights2019_df['TaxiOut'].fillna(flights2019_df['TaxiOut'].median(), inplace=True)
flights2019_df.isna().sum()

DayOfWeek                     0
FlightDate                    0
Reporting_Airline             0
Tail_Number                   0
Origin                        0
OriginCityName                0
OriginState                   0
Dest                          0
DestCityName                  0
DestState                     0
CRSDepTime                    0
DepTime                       0
DepDelay                      0
DepDel15                      0
DepartureDelayGroups          0
TaxiOut                       0
WheelsOff                133977
Cancelled                     0
CancellationCode        7284391
Distance                      0
DistanceGroup                 0
CarrierDelay            6030063
WeatherDelay            6030063
NASDelay                6030063
SecurityDelay           6030063
LateAircraftDelay       6030063
CRSArrTime                    0
ArrTime                       0
ArrDelay                      0
dtype: int64

In [26]:
# fill in WheelsOff with the CRSDepTime + the TaxiOut time

flights2019_df['WheelsOff'].fillna((flights2019_df['CRSDepTime'] + flights2019_df['TaxiOut']), inplace=True)
flights2019_df.isna().sum()

DayOfWeek                     0
FlightDate                    0
Reporting_Airline             0
Tail_Number                   0
Origin                        0
OriginCityName                0
OriginState                   0
Dest                          0
DestCityName                  0
DestState                     0
CRSDepTime                    0
DepTime                       0
DepDelay                      0
DepDel15                      0
DepartureDelayGroups          0
TaxiOut                       0
WheelsOff                     0
Cancelled                     0
CancellationCode        7284391
Distance                      0
DistanceGroup                 0
CarrierDelay            6030063
WeatherDelay            6030063
NASDelay                6030063
SecurityDelay           6030063
LateAircraftDelay       6030063
CRSArrTime                    0
ArrTime                       0
ArrDelay                      0
dtype: int64

In [27]:
# check the shapes of the whole dataframe as well as just the cancelled flights

display(flights2019_df.shape)
display(flights2019_df[flights2019_df['Cancelled'] == 1].shape)

(7419316, 29)

(134925, 29)

134,925 cancelled + 7,284,391 NaN Cancelation Code = 7,419,316 total --> correct

In [28]:
# replace NaN with 0 for CancellationCode and the reasons for Delay

flights2019_df['CancellationCode'].fillna(0, inplace=True)
flights2019_df['CarrierDelay'].fillna(0, inplace=True)
flights2019_df['WeatherDelay'].fillna(0, inplace=True)
flights2019_df['NASDelay'].fillna(0, inplace=True)
flights2019_df['SecurityDelay'].fillna(0, inplace=True)
flights2019_df['LateAircraftDelay'].fillna(0, inplace=True)
flights2019_df.isna().sum()

DayOfWeek               0
FlightDate              0
Reporting_Airline       0
Tail_Number             0
Origin                  0
OriginCityName          0
OriginState             0
Dest                    0
DestCityName            0
DestState               0
CRSDepTime              0
DepTime                 0
DepDelay                0
DepDel15                0
DepartureDelayGroups    0
TaxiOut                 0
WheelsOff               0
Cancelled               0
CancellationCode        0
Distance                0
DistanceGroup           0
CarrierDelay            0
WeatherDelay            0
NASDelay                0
SecurityDelay           0
LateAircraftDelay       0
CRSArrTime              0
ArrTime                 0
ArrDelay                0
dtype: int64

In [29]:
# add a boolean column for whether a flight is delayed or not (based on whether it arrives late)

flights2019_df['Delayed'] = 0

flights2019_df.loc[flights2019_df['ArrDelay'] <= 0, 'Delayed'] = 0
flights2019_df.loc[flights2019_df['ArrDelay'] > 0, 'Delayed'] = 1

flights2019_df['Delayed'].value_counts()

0    4905121
1    2514195
Name: Delayed, dtype: int64

In [30]:
# display the clean dataframe

flights2019_df

DayOfWeek  FlightDate Reporting_Airline Tail_Number Origin  \
0                4  2019-08-01                DL      N354NW    ATL   
1                4  2019-08-01                DL      N320US    DFW   
2                4  2019-08-01                DL      N931DN    IAH   
3                4  2019-08-01                DL      N851DN    PDX   
4                4  2019-08-01                DL      N775DE    SLC   
...            ...         ...               ...         ...    ...   
7422032          5  2019-11-08                DL      N978AT    HOU   
7422033          5  2019-11-08                DL      N6716C    CHS   
7422034          5  2019-11-08                DL      N851DN    MSP   
7422035          5  2019-11-08                DL      N352NB    GEG   
7422036          5  2019-11-08                DL      N352NB    SLC   

                OriginCityName OriginState Dest           DestCityName  \
0                  Atlanta, GA          GA  DFW  Dallas/Fort Worth, TX   
1        Dallas/Fort Worth, TX          TX  ATL            Atlanta, GA   
2                  Houston, TX          TX  ATL            Atlanta, GA   
3                 Portland, OR          OR  SLC     Salt Lake City, UT   
4           Salt Lake City, UT          UT  PDX           Portland, OR   
...                        ...         ...  ...                    ...   
7422032            Houston, TX          TX  ATL            Atlanta, GA   
7422033         Charleston, SC          SC  ATL            Atlanta, GA   
7422034        Minneapolis, MN          MN  BWI          Baltimore, MD   
7422035            Spokane, WA          WA  SLC     Salt Lake City, UT   
7422036     Salt Lake City, UT          UT  GEG            Spokane, WA   

        DestState  CRSDepTime  DepTime  DepDelay  DepDel15  \
0              TX        1954   2025.0      31.0       1.0   
1              GA        1709   1706.0      -3.0       0.0   
2              GA        1749   1829.0      40.0       1.0   
3              UT        1310   1306.0      -4.0       0.0   
4              OR         829    820.0      -9.0       0.0   
...           ...         ...      ...       ...       ...   
7422032        GA        1324   1320.0      -4.0       0.0   
7422033        GA         700    719.0      19.0       1.0   
7422034        MD        1800   1754.0      -6.0       0.0   
7422035        UT        1015   1009.0      -6.0       0.0   
7422036        WA         844    836.0      -8.0       0.0   

         DepartureDelayGroups  TaxiOut  WheelsOff  Cancelled CancellationCode  \
0                         2.0     17.0     2042.0          0                0   
1                        -1.0     19.0     1725.0          0                0   
2                         2.0     12.0     1841.0          0                0   
3                        -1.0     13.0     1319.0          0                0   
4                        -1.0     16.0      836.0          0                0   
...                       ...      ...        ...        ...              ...   
7422032                  -1.0     14.0     1334.0          0                0   
7422033                   1.0     10.0      729.0          0                0   
7422034                  -1.0     22.0     1816.0          0                0   
7422035                  -1.0     11.0     1020.0          0                0   
7422036                  -1.0     14.0      850.0          0                0   

         Distance  DistanceGroup  CarrierDelay  WeatherDelay  NASDelay  \
0             731              3           1.0           0.0       0.0   
1             731              3           0.0           0.0       0.0   
2             689              3          17.0           0.0       0.0   
3             630              3           0.0           0.0       0.0   
4             630              3           0.0           0.0       0.0   
...           ...            ...           ...           ...       ...   
7422032       696              

In [31]:
# export the clean dataframe with no NaN values to a csv

flights2019_df.to_csv('flights2019clean.csv')